### ISSUE TITLE
Running logical_clock3.tla turns my laptop into a heater

### ISSUE THREAD

**Issue by ToddG** [tlaplus/tlaplus#446]
**CONTEXT**
I'm writing up my experiences with writing a simplified logical clock spec here:

https://github.com/ToddG/experimental/tree/master/tla/eodc/03-logical-clocks/tla_simplified_logical_clock

At the bottom of this readme, I'm trying to run the logical_clock3.tla spec using the toolbox (screenshots attached). However, the spec does not complete, despite my having set it to the most minimal config that takes < 1 sec using `tlc`.

This commit includes the Models that do not want to complete on my machine...
https://github.com/ToddG/experimental/commit/6a77fc8555285e9442429e2e7d2b46d3fcdbc4d8

**REPRO**
See screenshots and commit above for the Model. Attempting to run this spec:

https://github.com/ToddG/experimental/blob/master/tla/eodc/03-logical-clocks/tla_simplified_logical_clock/logical_clock3.tla

**ERROR**
This did not complete even after 5 or so minutes.

Perhaps this is a configuration issue? I'm running everything else using the command line and using simple `.cfg` files like this one:

https://github.com/ToddG/experimental/blob/master/tla/eodc/03-logical-clocks/tla_simplified_logical_clock/A.cfg

**Comment by ToddG** [comment]
If there's a FAQ for reporting bugs that specifies how you want these reported and what information you want/need included, please LKM.

**Comment by lemmy** [comment]
The Toolbox generates a `MC.tla` and `MC.cfg` in `logical_clock3.toolbox/Model_1/` that you should compare to your hand-written one (might as well run command-line TLC on it).

**Comment by lemmy** [comment]
By the way, parsing TLC's output is simple enough to not require xml/json.  Take inspiration from:

```bash
#!/bin/bash
BRANCH="master"

for f in out*;
do
    
  ## Munge report data
  THROUGHPUT=$(cat ${SPEC}/$f | sed -n -e 's/^Progress(.*) at .* states generated (\(.*\) s\/min),.*/\1/p' | sed -e 's/,//g' | tail -1)
  RUNTIME=$(cat ${SPEC}/$f | sed -n -e 's/^Finished in \(.*\)ms at (.*)/\1/p')
  REVISION=$(cat ${SPEC}/$f | sed -n -e 's/^TLC2 Version .* (rev: \(.*\))/\1/p')
  BUILD_DATE=$(date +%s)

  if [[ -n $THROUGHPUT && -n $RUNTIME ]]; then
    echo ${BUILD_NUMBER},$BUILD_DATE,$REVISION,$BRANCH,${CLOUD},${SPEC},"Throughput",$THROUGHPUT >> ${JENKINS_HOME}/measurements.csv
    echo ${BUILD_NUMBER},$BUILD_DATE,$REVISION,$BRANCH,${CLOUD},${SPEC},"Runtime",$RUNTIME >> ${JENKINS_HOME}/measurements.csv
  fi
done
```

or

https://github.com/tlaplus/tlaplus/blob/master/general/performance/measure.sh
https://github.com/tlaplus/tlaplus/blob/master/general/performance/csv.sh

**Comment by ToddG** [comment]
Ah, cool. Thx! I'm not afraid of sed and awk...it's just after using `jq` and other such filters, I'm spoiled.

Ok, added the screenshot to my page. Spec Options just has the ActionConstraints in ACTION CONSTRAINT box. ```ActionConstraints == TRUE``` in this spec.

MC.cfg

```
CONSTANT
MaxClockValue <- const_15876797507583000
```

These are integer values in `A.cfg` for example:

```
CONSTANT MaxClockValue = 1
```

And the screenshot shows that this var should be an int, not a model value.

Q: Can you link me either a FAQ or the java src that describes the toolbox directory and file structures?

**Comment by lemmy** [comment]
Since you mention code, https://github.com/tlaplus/tlaplus/issues/447 might be of interest.  However, don't waste your time on the code.  Once you are familiar with TLC, you will realize that the .toolbox directory is just Eclipse metadata and TLC files.

**Comment by lemmy** [comment]
The Toolbox model has no state constraint (CONSTRAINT) while your hand-written .cfg has one.

**Comment by ToddG** [comment]
Ah! Ok, thx! I will retry shortly.

-------- Original Message --------

**Comment by ToddG** [comment]
Thanks for the help! 

I've verified that adding 'Constraints' to the CONSTRAINT box fixed things, and updated my readme here https://github.com/ToddG/experimental/tree/master/tla/eodc/03-logical-clocks/tla_simplified_logical_clock

I'm not sure why the IDE would let you run with an invalid configuration (assuming this is an invalid configuration). In any event, I'm unblocked.

**Comment by lemmy** [comment]
What makes you believe that it is an invalid configuration?

**Comment by ToddG** [comment]
Oh, I don't know if it is. I don't know the tool and it's use cases well enough to state that. 

I can just say that from the perspective of a new user, launching a model that doesn't complete b/c it is missing an element of the configuration (in this case the CONSTRAINT)... without any feedback... well it's unexpected.

**Comment by lemmy** [comment]
If you check a model with an infinite state space, TLC will run forever (regardless of whether TLC was launched from the command-line or the Toolbox).

https://en.wikipedia.org/wiki/Halting_problem

**Comment by ToddG** [comment]
Sort of like:
```
while TRUE
    wait() /* or whatever*/
```
You know much more about this space than I do. I'm just approaching this from the dev-on-the street perspective... If I were to launch a server, I'd typically see output like `server starting...` and whatever. From the context, I'd know that the server should be long running and I wouldn't expect it to complete. Similarly, if I were to run something that's supposed to be quick, like a compile step or printing something, anything that's supposed to complete... and this process _knows_ it cannot complete, then I'd expect at least a log message stating we are entering infinity. I guess I adhere to the principle of least surprise [Raymond, The Art of Unix Programming]... though my background may be different from the typical tla+ target audience.

**Comment by lemmy** [comment]
Background information: https://link.springer.com/chapter/10.1007/978-3-642-35746-6_1

**Comment by lemmy** [comment]
Please reopen if anything left undone.

**Comment by ToddG** [comment]
Ordered a copy of the Springer book, thx for the recommendation.

**Comment by ToddG** [comment]
Markus,
Thought you might be interested in this. I ordered the book you recommended from Springer. Here's what has happened so far:

1. Ordered and paid for the book you linked to above (Tools for Practical Software Verification)
2. Springer shipped me some random book, I forget what it was
3. I contacted customer support, and after many emails, finally convinced them they made a mistake
4. I returned the book they sent me
5. Weeks later, I had to contact customer support and convince them to send me the correct book
6. Today I received a notice from a collections agency b/c apparently they invoiced me for the book (again)

I'm still digging into this, but I rue the day I ordered this book. I will never, ever, do business with Springer again.
